luarules/gadgets/unit_prevent_cloaked_unit_reclaim.lua

```diff
@@ -0,0 +1,68 @@
+function gadget:GetInfo()
+    return {
+        name      = "Prevent Cloaked Unit Reclaim",
+        desc      = "Prevents builders from reclaiming cloaked units",
+        author    = "hihoman23",
+        date      = "Apr 2024",
+        license   = "GNU GPL, v2 or later",
+        layer     = 0,
+        enabled   = true  --  loaded by default?
+    }
+end
+
+local builders = {}
+
+
+if gadgetHandler:IsSyncedCode() then
+    local GetUnitAllyTeam = Spring.GetUnitAllyTeam
+    local GetUnitIsCloaked = Spring.GetUnitIsCloaked
+    local GetUnitWorkerTask = Spring.GetUnitWorkerTask
+    local GiveOrderToUnit = Spring.GiveOrderToUnit
+    local GetUnitCurrentCommand = Spring.GetUnitCurrentCommand
+    local GetAllUnits = Spring.GetAllUnits
+    local GetUnitDefID = Spring.GetUnitDefID
+    local IsUnitInRadar = Spring.IsUnitInRadar
+
+
+    function gadget:AllowCommand(unitID, unitDefID, unitTeam, cmdID, cmdParams)
+        if (cmdID == CMD.RECLAIM) and (#cmdParams == 1) and GetUnitIsCloaked(cmdParams[1]) and (GetUnitAllyTeam(unitID) ~= GetUnitAllyTeam(cmdParams[1])) and not IsUnitInRadar(cmdParams[1], GetUnitAllyTeam(unitID)) then
+            return false
+        end
+        return true
+    end
+    
+    function gadget:AllowUnitCloak(unitID) -- cancel reclaim commands
+        for bID, _ in pairs(builders) do
+            local cmd, target = GetUnitWorkerTask(bID)
+            if cmd == CMD.RECLAIM and target == unitID and (GetUnitAllyTeam(bID) ~= GetUnitAllyTeam(unitID)) and not IsUnitInRadar(unitID, GetUnitAllyTeam(bID)) then
+                local _, _, cmdTag = GetUnitCurrentCommand(bID, 1)
+                GiveOrderToUnit(bID, CMD.REMOVE, {cmdTag}, {})
+            end
+        end
+        return true
+    end
```

Comment:
Iterating over every builder in the game could potentially get *very* expensive. This only occurring when a unit cloaks will help, but this can be optimized further. Instead of caching every single builder, I recommend doing `Spring.GetUnitsInCylinder` and using the maximum buildrange of all cached units as the query size. This means you'll also need to cache buildrange, which can vary based on which units are enabled. I recommend something like this

```lua
-- on init
local maxBuildDist = 0
local canReclaim = {}
for unitDefID, unitDef in pairs(UnitDefs) do
	if unitDef.canReclaim then
		canReclaim[unitDefID] = true
        if unitDef.buildDistance > maxBuildDist then
            maxBuildDist = unitDef.buildDistance
        end
	end
end

function gadget:AllowUnitCloak(unitID) -- cancel reclaim commands
    local x, y, z = Spring.GetUnitPosition(unitID)
    local units = Spring.GetUnitsInCylinder(x, z, maxBuildDist)
    for uid, _ in pairs(units) do
        local unitDefID = Spring.GetUnitDefID(unitID)
        if canReclaim[unitDefID] then
            -- same order stuff as you have
        end
    end
    return true
end
```
Note this is pseudocode I wrote without testing, might need a little tweaking